### PR TITLE
Add a cleanLeaderPrefix function to clean up stale leader entries in core/leader

### LINKED ISSUE #679: excessive vault/core/leader entries
I reinstalled vault (0.3.1) last night from scratch.  Consul (0.5.2) backend.

I already have 1,730 entries in vault/core/leader in Consul.

I just reinstalled from scratch because I had 2.8M leases from the 0.2.0 behavior of creating a lease for every read of any generic secret, so I'm a bit wary...

### PR DESCRIPTION
**Pull request by jefferai**
Fixes #679.

### CODE REVIEW

File: vault/core.go [hunk 1]
```diff
@@ -1610,8 +1616,35 @@ func (c *Core) advertiseLeader(uuid string) error {
...
 }
 
+func (c *Core) cleanLeaderPrefix(uuid string) {
+	keys, err := c.barrier.List(coreLeaderPrefix)
+	if err != nil {
+		c.logger.Printf("[ERR] core: failed to list entries in core/leader: %v", err)
+		return
+	}
+	if len(keys) == 0 {
+		c.logger.Print("[ERR] core: found no entries under core/leader (should have found our own)")
```
Comment (armon) [inline vault/core.go]: There is a race condition, where this is possible. No need to log an error

File: vault/core.go [hunk 2]
```diff
@@ -239,6 +239,10 @@ type Core struct {
 	// metricsCh is used to stop the metrics streaming
 	metricsCh chan struct{}
 
+	// leaderPrefixCleanupCh is used to stop
+	// the leader prefix cleansing operation
+	leaderPrefixCleanupCh chan struct{}
```
Comment (armon) [inline vault/core.go]: Just thread through the `leaderCh` this is not needed!

File: vault/core.go [hunk 3]
```diff
@@ -1603,6 +1607,8 @@ func (c *Core) acquireLock(lock physical.Lock, stopCh <-chan struct{}) <-chan st
 
 // advertiseLeader is used to advertise the current node as leader
 func (c *Core) advertiseLeader(uuid string) error {
```
Comment (armon) [inline vault/core.go]: This function can take the leaderCh from the caller

File: vault/core.go [hunk 4]
```diff
@@ -1610,8 +1616,35 @@ func (c *Core) advertiseLeader(uuid string) error {
...
+	keys, err := c.barrier.List(coreLeaderPrefix)
+	if err != nil {
+		c.logger.Printf("[ERR] core: failed to list entries in core/leader: %v", err)
+		return
+	}
+	if len(keys) == 0 {
+		c.logger.Print("[ERR] core: found no entries under core/leader (should have found our own)")
+		return
+	}
+	for {
```
Comment (armon) [inline vault/core.go]: Simpler to do `for len(keys) > 0` to remove check above and within the loop

File: vault/core.go [hunk 5]
```diff
@@ -1610,8 +1616,35 @@ func (c *Core) advertiseLeader(uuid string) error {
...
+		select {
+		case <-time.After(time.Second):
```
Comment (armon) [inline vault/core.go]: This is a pretty slow pace. I'd also move this to a `const`

File: vault/core.go (hunk 2 above)
Comment (jefferai) [inline vault/core.go]: Good point!

**Pull request merged**